build-setup/vscode/using_vscode.md

```diff
@@ -0,0 +1,232 @@
+# Using Visual Studio Code with `ml-cpp`
+
+Previously, we explained how to [set up CLion for `ml-cpp`](../clion/using_clion.md). You may find, that CLion has some
+performance problems when indexing `ml-cpp` project: it may take very long to open the project or CLion may freeze and
+fail. In this case, [Visual Studio Code](https://code.visualstudio.com/) can be an alternative. When combined with
+`clangd`, it is significantly faster, while still being able to provide features like autocompletion, code navigation,
+static code analysis, etc.
+
+## Prerequisites
+
+In this tutorial, we assume that you have `gcc` or `clang` installed as described in the
+[CLion tutorial](../clion/using_clion.md). We also assume that you already created `compile_commands.json` file as
+described there since `clangd` also uses it.
+
+[Install `clangd`](https://clangd.llvm.org/installation.html). Version 11.0 had a significant performance improvement,
+so it is worse to go an extra mile and install the latest stable version. Install
+[Visual Studio Code](https://code.visualstudio.com/) with
+[clangd plugin](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-clangd).
+
+In **Settings**, make sure *C_Cpp:Autocomplete* and *C_Cpp:Intelli Sense Engine* from the native C++ plugin of VS Code
+are disabled. Note that we don't want to disable the complete C/C++ Extension, since we are still going to use the
+debugging capabilities.
+
+Now for the `clangd` plugin, make sure that you have the correct path specified in **Settings** and *Clangd: Path*.
+*Clangd: Arguments* can look something like this:
+
+```text
+-log=verbose
+-pretty
+--header-insertion=iwyu
+--suggest-missing-includes
+-j=4
+--all-scopes-completion
+--background-index=0
+--clang-tidy
+--compile-commands-dir=./
+```
+
+## User tasks
+
+You can use [user tasks](https://code.visualstudio.com/docs/editor/tasks) to integrate VS Code with external tools. For
+example, re-format the project using a the docker container with correct clang-format version or build certain parts of
+the project in Debug mode. Later, we will see how we can use these tasks as dependencies for debugging.
+
+Here is an example of user tasks specified in `tasks.json`:
+
+```json
+{
+    "version": "2.0.0",
+    "tasks": [
+        {
+            "label": "Format code",
+            "type": "shell",
+            "command": "docker run --rm -v ${workspaceFolder}:/ml-cpp -u $(id -u):$(id -g) docker.elastic.co/ml-dev/ml-check-style:2 /ml-cpp/dev-tools/clang-format.sh",
```

Comment:
>  -u $(id -u):$(id -g)

I'm not sure what should be passed for the login here. Could we invoke `clang-format.sh` directly outside of docker assuming it is installed locally

Reply:
You require to use `clang-format` version 5.0.1. I am using docker container here to ensure the correct version and to avoid mingling with host dependencies.

Reply:
Once I logged into docker.elastic.co I could run this in a shell.

I changed the command to call `run_docker_clang_format.sh` because if the docker image is updated the shell script should also be changed. Hardcoding the version here means it will go out of date eventually.